### ISSUE TITLE
-gseparate-dwarf reads like it shoud work like -gsplit-dwarf(?), but doesn't

### ISSUE THREAD

**Issue by juj** [emscripten-core/emscripten#17814]
After talking with @dschuff about our Wasm debugging troubles, he suggested using the `-gseparate-dwarf` option to create separate [DWARF object files (.dwo)](https://gcc.gnu.org/wiki/DebugFission), with the impression that it would work like `-gsplit-dwarf` build option.

The documentation reads a bit similar:

```
``-gseparate-dwarf[=FILENAME]``
  [same as -g3 if passed at compile time, otherwise applies at link]
  Preserve debug information, but in a separate file on the side. This is the
  same as ``-g``, but the main file will contain no debug info. Instead, debug
  info will be present in a file on the side, in ``FILENAME`` if provided,
  otherwise the same as the wasm file but with suffix ``.debug.wasm``. While
  the main file contains no debug info, it does contain a URL to where the
  debug file is, so that devtools can find it. You can use
  ``-sSEPARATE_DWARF_URL=URL`` to customize that location (this is useful if
  you want to host it on a different server, for example).
```

If I have files
`a.cpp`
```cpp
int foo(int a, int b)
{
	return a + b;
}
```
`b.cpp`
```cpp
#include <stdio.h>

int foo(int a, int b);

int main()
{
	printf("%d\n", foo(1, 2));
}
```

From the doc, I get the idea that `emcc a.cpp b.cpp -o test.js -gseparate-dwarf` would produce:
 - `test.wasm`: contains the compiled output code with no debug info
 - `test.wasm.debug.wasm`: contains the debug info sections for `test.wasm`

But this is not the case. The file `test.wasm` indeed does not contain DWARF info (though it does still contain the Wasm `NAME` section for function names), but the file `test.wasm.debug.wasm` is not just a debug info container, but it is a full copy of the program, plus with the DWARF debugging info intact.

I.e. it seems that `emcc a.cpp b.cpp -o test.html -gseparate-dwarf` is essentially an alias to writing the two commands

`emcc a.cpp b.cpp -o test.js -g2`
`emcc a.cpp b.cpp -o test.wasm.debug.js -g3`

Is this intentional?

<hr>

A sidenote thought: the naming `test.wasm.debug.wasm` is maybe not ideal? Could this instead be renamed to `test.debug.wasm`?

<hr>

Then a second observation: the LLVM `-gsplit-dwarf` option does not seem to work with Wasm. The [original GCC docs state that](https://gcc.gnu.org/wiki/DebugFission)

" Use the -gsplit-dwarf option to enable the generation of split DWARF at compile time. This option must be used in conjunction with -c; Fission cannot be used when compiling and linking in the same step. "

i.e. my understanding here is that natively one would run

```
g++ -c a.cpp -o a.o -gsplit-dwarf
g++ -c b.cpp -o b.o -gsplit-dwarf
g++ a.o b.o -o test.exe
```
to produce an executable with separate `a.dwo` and `b.dwo` debug info files? (similar to how Microsoft's separate PDB files work?)

It seems that the `-gsplit-dwarf` option does not work with Emscripten: at `-c` compile commands it is getting ignored, but at link stage, it does generate a `.dwo` file, which is basically an empty Wasm object file:
```wast
(module
 ;; custom section "producers", size 112
)
```

And the `-gseparate-dwarf` option does not currently also work to separate/split DWARF debug info into distinct `.dwo`/`.wasm` files: rather it can only be used to generate a copy of the full output program with DWARF intact?

<hr>

Is the above accurate? Should there exist a method to generate separate individual per-object debugging symbol files that Chrome Wasm DWARF debugger could consume?

The issue with Chrome debugger is that it chokes at Unity scale of DWARF output, hanging when the very large DWARF enabled .wasm files are fed to it - hence looking for a way to split the debug info into separate per-object DWARF files. With the current `-gseparate-dwarf`, it does not help Chrome debugger, since there is still one big Wasm code+DWARF file present.

Or maybe I missed some step? Thanks!

**Comment by xbcnn** [comment]
Interesting how to generate/use the separate DWARF files. Please keep me informed.

**Comment by mere-human** [comment]
FYI: https://developer.chrome.com/blog/faster-wasm-debugging/

**Comment by xbcnn** [comment]
> FYI: https://developer.chrome.com/blog/faster-wasm-debugging/

Nice!

**Comment by xbcnn** [comment]
Seems the options don't support `-O2`. I got errors when I just changed the `-O0` to `-O2`.
```
emcc: warning: running limited binaryen optimizations because DWARF info requested (or indirectly required) [-Wlimited-postlink-optimizations]
Fatal: unknown tag with low_pc DW_TAG_skeleton_unit
emcc: error: '/Users/yangxiaobing/emsdk/upstream/bin/wasm-opt --post-emscripten -O2 --low-memory-unused --zero-filled-memory --pass-arg=directize-initial-contents-immutable --strip-producers out/mandelbrot.wasm -o out/mandelbrot.wasm -g --mvp-features' failed (returned 1)
```

**Comment by dschuff** [comment]
Hi, sorry I missed this 2 weeks ago... and sorry things are still a bit rough :D

First, regarding `-gseparate-dwarf` vs `-gsplit-dwarf`, I must not have been clear about the difference, sorry about that. `-gseparate-dwarf` is like `-split-dwarf` when the latter is used with the `dwp` (or in our case, `llvm-dwp`) tool. `-gsplit-dwarf` does create a `.dwo` for each `.o` file. But if the user wants to archive the debug info for any purpose other than debugging on the build machine, it's common to then use `dwp` to package all the `.dwo` files into a single file (`.dwp` file). A debugger that understands split-dwarf should be able to use either the `.dwo` files or the `.dwp` file for debugging. `-gseparate-dwarf` is like the latter use case in that it results in a wasm file with just the code, and another wasm file with "just" the debug info.

(I say "just" because as you noticed it's not just dwarf; there are 2 issues: first it turns out that in order to interpret the DWARF info, you actually need the wasm sections that declare types function signatures, and a few others. and second, the chrome debugger extension actually has a bug that means it actually requires the code section to work as well. Once that fix is rolled out we can strip the code section out of the .debug.wasm file, but we'll always need the signature declarations).

It's a good point about having the name section left over in the main file after the separation. Some folks seem to want to keep that in non-dwarf binaries (since the engine understands it and will generate readable stack traces even when not debugging) but I think it probably does make sense to not have it by default.

And regarding gsplit-dwarf usage, https://developer.chrome.com/blog/faster-wasm-debugging/ is the best current documentation, thanks for posting that. We should definitely make `-gdwarf-5` and `-gpubnames` the default when using split-dwarf. Those should maybe be the default overall, but currently it's also still possible to use debug info with O2 and above (which means the Binaryen optimizer is running), and Binaryen doesn't support rewriting DWARF5. Binaryen also doesn't know how to rewrite dwo or dwp files, it only works when not splitting dwarf. But even when it does work, it has to inhibit some of its optimizations because of its limited understanding of DWARF.
So tl:dr if you want to do interactive debugging with DWARF you're best off using O0 or O1 (even with O1 there are some known issues with debug info quality, but @aheejin is working on that). 
The further future direction is a little less sure. We know that some users (especially game developers!) really like to debug *exactly* the super-optimized bits that they are shipping, but Binaryen isn't in a great place for that right now, and it would be a lot of work to make that work. We for sure intend to at least make Binaryen preserve line table information to get good stack traces.

**Comment by dschuff** [comment]
> > the chrome debugger extension actually has a bug that means it actually requires the code section to work as well. Once that fix is rolled out we can strip the code section out of the .debug.wasm file, but we'll always need the signature declarations
> 
> @dschuff I'd like to follow this bug so I know when I can strip the code section, is there a crbug or the like?

I don't recall specifically, but @pfaffe might know.

**Comment by pfaffe** [comment]
I don't think we have a crbug for this.
